RecoTauTag/RecoTau/src/RecoTauQualityCuts.cc

```diff
@@ -3,18 +3,40 @@
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
 #include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
 
 #include <boost/bind.hpp>
 
 namespace reco { namespace tau {
 
 namespace {
-  // Get the KF track if it exists.  Otherwise, see if PFCandidate has a GSF track.
-  const reco::TrackBaseRef getTrackRef(const PFCandidate& cand) 
+  const reco::Track* getTrack(const Candidate& cand) 
+  { 
+    const PFCandidate* pfCandPtr = dynamic_cast<const PFCandidate*>(&cand);
+    if (pfCandPtr) {
+      // Get the KF track if it exists.  Otherwise, see if PFCandidate has a GSF track.
+      if      ( pfCandPtr->trackRef().isNonnull()    ) return pfCandPtr->trackRef().get();
+      else if ( pfCandPtr->gsfTrackRef().isNonnull() ) return pfCandPtr->gsfTrackRef().get();
+      else return nullptr;
+    }
+
+    const pat::PackedCandidate* packedCand = dynamic_cast<const pat::PackedCandidate*>(&cand);
+    if (packedCand && packedCand->hasTrackDetails())
+    	return &packedCand->pseudoTrack();
+
+    return nullptr;
+  }
+
+  const reco::TrackBaseRef getTrackRef(const Candidate& cand) 
   {
-    if ( cand.trackRef().isNonnull() ) return reco::TrackBaseRef(cand.trackRef());
-    else if ( cand.gsfTrackRef().isNonnull() ) return reco::TrackBaseRef(cand.gsfTrackRef());
-    else return reco::TrackBaseRef();
+    const PFCandidate* pfCandPtr = dynamic_cast<const PFCandidate*>(&cand);
+    if (pfCandPtr) {
+      if      ( pfCandPtr->trackRef().isNonnull()    ) return reco::TrackBaseRef(pfCandPtr->trackRef());
+      else if ( pfCandPtr->gsfTrackRef().isNonnull() ) return reco::TrackBaseRef(pfCandPtr->gsfTrackRef());
```

Comment:
looking at the profiler, the compiler apparently does not optimize away the two calls to pfCandPtr->trackRef().

There is also an extra cost overhead from making a TrackBase. 
Considering that gsfTrack is available much less frequently, I think that it's better to make getTrackRef and getGsfTrackRef separately, each returning their native types.

in this case the implementation of this method becomes simply
```c++
if (pfCandPtr) return pfCandPtr->trackRef();
else return TrackRef();
```